File: dataset/ZeroShot/GEMINI_gemini-2.5-pro/Task156.c
#include <stdio.h>
#include <stdlib.h>

// Node structure for the binary tree
typedef struct Node {
    int data;
    struct Node* left;
    struct Node* right;
} Node;

// Utility function to create a new node
Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) {
        // Handle memory allocation failure
        fprintf(stderr, "Error: Memory allocation failed.\n");
        return NULL;
    }
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Utility function to free the tree memory to prevent leaks
void freeTree(Node* node) {
    if (node == NULL) {
        return;
    }
    freeTree(node->left);
    freeTree(node->right);
    free(node);
}

/**
 * Performs a preorder traversal of the binary tree.
 * Root -> Left -> Right
 * @param node The root node of the tree/subtree.
 */
void preorderTraversal(Node* node) {
    if (node == NULL) {
        return;
    }
    printf("%d ", node->data);
    preorderTraversal(node->left);
    preorderTraversal(node->right);
}

/**
 * Performs an inorder traversal of the binary tree.
 * Left -> Root -> Right
 * @param node The root node of the tree/subtree.
 */
void inorderTraversal(Node* node) {
    if (node == NULL) {
        return;
    }
    inorderTraversal(node->left);
    printf("%d ", node->data);
    inorderTraversal(node->right);
}

/**
 * Performs a postorder traversal of the binary tree.
 * Left -> Right -> Root
 * @param node The root node of the tree/subtree.
 */
void postorderTraversal(Node* node) {
    if (node == NULL) {
        return;
    }
    postorderTraversal(node->left);
    postorderTraversal(node->right);
    printf("%d ", node->data);
}


int main() {
    // Test Case 1: A complete binary tree
    printf("--- Test Case 1: Complete Binary Tree ---\n");
    Node* root1 = createNode(1);
    if(root1) {
        root1->left = createNode(2);
        root1->right = createNode(3);
        if(root1->left) {
            root1->left->left = createNode(4);
            root1->left->right = createNode(5);
        }
        if(root1->right) {
            root1->right->left = createNode(6);
            root1->right->right = createNode(7);
        }
    }
    printf("Preorder:  ");
    preorderTraversal(root1);
    printf("\n");
    printf("Inorder:   ");
    inorderTraversal(root1);
    printf("\n");
    printf("Postorder: ");
    postorderTraversal(root1);
    printf("\n\n");
    freeTree(root1);

    // Test Case 2: A right-skewed tree
    printf("--- Test Case 2: Right-Skewed Tree ---\n");
    Node* root2 = createNode(10);
    if(root2) {
        root2->right = createNode(20);
        if(root2->right) {
            root2->right->right = createNode(30);
            if(root2->right->right) {
                root2->right->right->right = createNode(40);
            }
        }
    }
    printf("Preorder:  ");
    preorderTraversal(root2);
    printf("\n");
    printf("Inorder:   ");
    inorderTraversal(root2);
    printf("\n");
    printf("Postorder: ");
    postorderTraversal(root2);
    printf("\n\n");
    freeTree(root2);
    
    // Test Case 3: A left-skewed tree
    printf("--- Test Case 3: Left-Skewed Tree ---\n");
    Node* root3 = createNode(50);
    if(root3) {
        root3->left = createNode(40);
        if(root3->left) {
            root3->left->left = createNode(30);
            if(root3->left->left) {
                root3->left->left->left = createNode(20);
            }
        }
    }
    printf("Preorder:  ");
    preorderTraversal(root3);
    printf("\n");
    printf("Inorder:   ");
    inorderTraversal(root3);
    printf("\n");
    printf("Postorder: ");
    postorderTraversal(root3);
    printf("\n\n");
    freeTree(root3);

    // Test Case 4: An empty tree (NULL root)
    printf("--- Test Case 4: Empty Tree ---\n");
    Node* root4 = NULL;
    printf("Preorder:  (empty)\n");
    printf("Inorder:   (empty)\n");
    printf("Postorder: (empty)\n\n");
    freeTree(root4);

    // Test Case 5: A tree with only one node
    printf("--- Test Case 5: Single Node Tree ---\n");
    Node* root5 = createNode(100);
    printf("Preorder:  ");
    preorderTraversal(root5);
    printf("\n");
    printf("Inorder:   ");
    inorderTraversal(root5);
    printf("\n");
    printf("Postorder: ");
    postorderTraversal(root5);
    printf("\n\n");
    freeTree(root5);

    return 0;
}

File: dataset/ZeroShot/GEMINI_gemini-2.5-pro/Task156.py
class Node:
    """Node class for the binary tree"""
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def preorder_traversal(node):
    """
    Performs a preorder traversal of the binary tree.
    Root -> Left -> Right
    :param node: The root node of the tree/subtree.
    """
    if node is None:
        return
    print(node.data, end=" ")
    preorder_traversal(node.left)
    preorder_traversal(node.right)

def inorder_traversal(node):
    """
    Performs an inorder traversal of the binary tree.
    Left -> Root -> Right
    :param node: The root node of the tree/subtree.
    """
    if node is None:
        return
    inorder_traversal(node.left)
    print(node.data, end=" ")
    inorder_traversal(node.right)

def postorder_traversal(node):
    """
    Performs a postorder traversal of the binary tree.
    Left -> Right -> Root
    :param node: The root node of the tree/subtree.
    """
    if node is None:
        return
    postorder_traversal(node.left)
    postorder_traversal(node.right)
    print(node.data, end=" ")

def main():
    # Test Case 1: A complete binary tree
    print("--- Test Case 1: Complete Binary Tree ---")
    root1 = Node(1)
    root1.left = Node(2)
    root1.right = Node(3)
    root1.left.left = Node(4)
    root1.left.right = Node(5)
    root1.right.left = Node(6)
    root1.right.right = Node(7)
    
    print("Preorder:  ", end="")
    preorder_traversal(root1)
    print()
    print("Inorder:   ", end="")
    inorder_traversal(root1)
    print()
    print("Postorder: ", end="")
    postorder_traversal(root1)
    print("\n")

    # Test Case 2: A right-skewed tree
    print("--- Test Case 2: Right-Skewed Tree ---")
    root2 = Node(10)
    root2.right = Node(20)
    root2.right.right = Node(30)
    root2.right.right.right = Node(40)

    print("Preorder:  ", end="")
    preorder_traversal(root2)
    print()
    print("Inorder:   ", end="")
    inorder_traversal(root2)
    print()
    print("Postorder: ", end="")
    postorder_traversal(root2)
    print("\n")

    # Test Case 3: A left-skewed tree
    print("--- Test Case 3: Left-Skewed Tree ---")
    root3 = Node(50)
    root3.left = Node(40)
    root3.left.left = Node(30)
    root3.left.left.left = Node(20)

    print("Preorder:  ", end="")
    preorder_traversal(root3)
    print()
    print("Inorder:   ", end="")
    inorder_traversal(root3)
    print()
    print("Postorder: ", end="")
    postorder_traversal(root3)
    print("\n")
    
    # Test Case 4: An empty tree (None root)
    print("--- Test Case 4: Empty Tree ---")
    root4 = None
    
    print("Preorder:  ", end="")
    preorder_traversal(root4)
    print("(empty)")
    print("Inorder:   ", end="")
    inorder_traversal(root4)
    print("(empty)")
    print("Postorder: ", end="")
    postorder_traversal(root4)
    print("(empty)\n")

    # Test Case 5: A tree with only one node
    print("--- Test Case 5: Single Node Tree ---")
    root5 = Node(100)

    print("Preorder:  ", end="")
    preorder_traversal(root5)
    print()
    print("Inorder:   ", end="")
    inorder_traversal(root5)
    print()
    print("Postorder: ", end="")
    postorder_traversal(root5)
    print("\n")

if __name__ == "__main__":
    main()